test/models/user_test.rb

```
@@ -0,0 +1,102 @@
+  end
+
+  test 'email should be present' do
+    user = build(:user, email: '     ')
+    assert_predicate user, :invalid?
+    assert_equal ({ email: ["can't be blank", 'is invalid'] }), user.errors.messages
+  end
+
+  test 'name maximum valid length' do
+    user = build(:user, name: 'a' * 50)
+    assert_predicate user, :valid?
+  end
+
+  test 'name exceeds maximum length' do
+    user = build(:user, name: 'a' * 51)
+    assert_predicate user, :invalid?
+  end
+
+  test 'email maximum valid total length' do
+    user = build(:user, email: 'a' * 64 + '@' + 'b' * 186 + '.com')
+    assert_predicate user, :valid?
+  end
+
+  test 'email exceeds maximum total length' do
+    user = build(:user, email: 'a' * 64 + '@' + 'b' * 187 + '.com')
+    assert_predicate user, :invalid?
+  end
+
+  test 'email exceeds maximum length of local part' do
+    user = build(:user, email: 'a' * 65 + '@example.com')
+    assert_predicate user, :invalid?
+  end
+
+  test 'email validation should accept valid addresses' do
+    user = build(:user)
+    valid_addresses = %w(
+      user@example.com
+      USER@foo.COM
+      A_US-ER@foo.bar.org
+      first.last@foo.jp
+      alice+bob@baz.cn
+    )
+    valid_addresses.each do |valid_address|
+      user.email = valid_address
+      assert user.valid?, "#{valid_address.inspect} should be valid"
+    end
+  end
+
+  test 'email validation should reject invalid addresses' do
+    user = build(:user)
+    invalid_addresses = %w(
+      plainaddress
+      #@%^%#$@#$@#.com
+      @example.com
+      Joe Smith <email@example.com>
+      email.example.com
+      email@example@example.com
+    )
+    invalid_addresses.each do |invalid_address|
+      user.email = invalid_address
+      assert user.invalid?, "#{invalid_address.inspect} should be invalid"
+    end
+  end
+
+  test 'email addresses should be unique' do
+    user1 = build(:user)
+    user2 = build(:user)
+    user1.email = user2.email.upcase
+    user1.save
```

Yes a system failure can occur. In this case it would silently return false if there were a system failure, or if somehow the user you are trying to create is invalid.

In rails there is both `save` and `save!`, the latter will raise an exception if the save fails. Consider using that instead wherever you are not explicitly checking the return value. The same applies for `create` and `create!`.